File: C++/Tree/100_Same Tree (Microsoft googgle).cpp
//////////////////////// DFS ////////////////

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if(p==NULL && q==NULL){
            return true;
        }
        if(p==NULL or q==NULL){
            return false;
        }
        
        if(p->val != q->val){
            return false;
        }
        
        return isSameTree(p->left, q->left) && isSameTree(p->right,q->right);
    }
};



/////////////////////// BFS //////////////////

class Solution {
public:
    
     bool isSame(TreeNode* p, TreeNode* q){
          if(p==NULL && q==NULL){
            return true;
        }
        if(p==NULL or q==NULL){
            return false;
        }
        
        if(p->val != q->val){
            return false;
        }
         return true;
     }
    
    bool isSameTree(TreeNode* p, TreeNode* q) {
        queue<TreeNode*> qp;
        queue<TreeNode*> qq;
        
        qp.push(p);
        qq.push(q);
        
        while(!qp.empty() && !qq.empty()){
            TreeNode* node1 = qp.front();
            qp.pop();
            TreeNode* node2 = qq.front();
            qq.pop();
            
            if(!isSame(node1, node2)){
                return false;
            }
            
            if(node1 != NULL){
                qp.push(node1->left);
                qq.push(node2->left);
                qp.push(node1->right);
                qq.push(node2->right);
            }
            
        }
        
        return true;

        
            }
};

File: Python/Tree/100_Same Tree (Microsoft googgle).py
#################### DFS ###################

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:

        
        def dfs(p,q):
            if not p and not q:
                return True
            if p and not q:
                return False
            if not p and q:
                return False
            
            if(q.val != p.val):
                return False
            ans = dfs(p.left, q.left) and dfs(p.right, q.right)    
            
            
            return ans
        
        return dfs(p,q)






################## BFS ############################
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:

        
        def dfs(p,q):
            if not p and not q:
                return True
            if not p or not q:
                return False           
            if(q.val != p.val):
                return False
            return True
        
        deq = deque([(p,q),])
        
        while deq:
            pp, qq = deq.popleft()
            if not dfs(pp, qq):
                return False
            if pp:
                deq.append((pp.left, qq.left))
                deq.append((pp.right, qq.right))
        
        return True
         